packages/teams-js/src/internal/appHelpers.ts

```diff
@@ -70,8 +86,43 @@ export function notifyFailureHelper(apiVersiontag: string, appInitializationFail
   ]);
 }
 
-export function notifySuccessHelper(apiVersionTag: string): void {
-  sendMessageToParent(apiVersionTag, app.Messages.Success, [version]);
+export async function notifySuccessHelper(apiVersionTag: string): Promise<NotifySuccessResponse> {
+  // The following implementation ensures that notify success can be called before the initialize
+  // call resolves completely, while still accessing the initialized runtime object without
+  // any issue.
+
+  // If the initialize already completed, dispatch notify success
+  if (GlobalVars.initializeCompleted) {
+    return callNotifySuccessInHost(apiVersionTag);
+  }
+
+  // If initialize hasn't been called yet, throw an error to the dev as the app hasn't initialized yet
+  if (!GlobalVars.initializePromise) {
+    throw new Error(errorLibraryNotInitialized);
+  }
+
+  // If initialize is still waiting for response, dispatch the call after initialize
+  // finishes to have the full runtime object instantiated.
+  return GlobalVars.initializePromise.then(() => callNotifySuccessInHost(apiVersionTag));
```

Comment:
Don't we want to handle if the initializePromise is rejected?

Reply:
Good question! If you look at `communication.initializeCommunication`, if we never receive a response to our initialize request we never change the value of `Communication.parentOrigin` from null to the host's actual origin, which will prevent the notifySuccess message from being sent to the host at all (which you can see in `communication.sendRequestToTargetWindowHelper`)